controllers/capabilities/activedoc_threescale_reconciler.go

```diff
@@ -0,0 +1,302 @@
+)
+
+type ActiveDocThreescaleReconciler struct {
+	*reconcilers.BaseReconciler
+	resource            *capabilitiesv1beta1.ActiveDoc
+	threescaleAPIClient *threescaleapi.ThreeScaleClient
+	providerAccountHost string
+	logger              logr.Logger
+}
+
+func NewActiveDocThreescaleReconciler(b *reconcilers.BaseReconciler, resource *capabilitiesv1beta1.ActiveDoc, threescaleAPIClient *threescaleapi.ThreeScaleClient, providerAccountHost string, logger logr.Logger) *ActiveDocThreescaleReconciler {
+	return &ActiveDocThreescaleReconciler{
+		BaseReconciler:      b,
+		resource:            resource,
+		threescaleAPIClient: threescaleAPIClient,
+		providerAccountHost: providerAccountHost,
+		logger:              logger.WithValues("3scale Reconciler", providerAccountHost),
+	}
+}
+
+func (s *ActiveDocThreescaleReconciler) Reconcile() (*threescaleapi.ActiveDoc, error) {
+	s.logger.V(1).Info("START")
+
+	remoteActiveDocs, err := s.threescaleAPIClient.ListActiveDocs()
+	if err != nil {
+		return nil, err
+	}
+
+	var remoteActiveDoc *threescaleapi.ActiveDoc
+
+	for idx := range remoteActiveDocs.ActiveDocs {
+		// s.resource.Spec.SystemName is not nil (defaults are set)
+		if *remoteActiveDocs.ActiveDocs[idx].Element.SystemName == *s.resource.Spec.SystemName {
+			remoteActiveDoc = &remoteActiveDocs.ActiveDocs[idx]
+			break
+		}
+	}
+
+	desiredProductID, err := s.getDesiredProductIDFromCR()
+	if err != nil {
+		return nil, err
+	}
+
+	desiredOpenapiObj, err := s.getDesiredActiveDocBody()
+	if err != nil {
+		return nil, err
+	}
+
+	desiredBodyRaw, err := desiredOpenapiObj.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	desiredBody := string(desiredBodyRaw)
+
+	if remoteActiveDoc == nil {
+		newActiveDoc := &threescaleapi.ActiveDoc{
+			Element: threescaleapi.ActiveDocItem{
+				Name:                   &s.resource.Spec.Name,
+				SystemName:             s.resource.Spec.SystemName,
+				Body:                   &desiredBody,
+				Description:            s.resource.Spec.Description,
+				Published:              s.resource.Spec.Published,
+				SkipSwaggerValidations: s.resource.Spec.SkipSwaggerValidations,
+				ServiceID:              desiredProductID,
+			},
+		}
+
+		remoteActiveDoc, err = s.threescaleAPIClient.CreateActiveDoc(newActiveDoc)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	update := false
+	updatedActiveDoc := &threescaleapi.ActiveDoc{
+		Element: threescaleapi.ActiveDocItem{
+			ID: remoteActiveDoc.Element.ID,
+		},
+	}
+
+	if remoteActiveDoc.Element.Name == nil || *remoteActiveDoc.Element.Name != s.resource.Spec.Name {
+		s.logger.V(1).Info("update Name", "Difference", cmp.Diff(remoteActiveDoc.Element.Name, s.resource.Spec.Name))
+		updatedActiveDoc.Element.Name = &s.resource.Spec.Name
+		update = true
+	}
+
+	// s.resource.Spec.SystemName is not nil (defaults are set)
+	if remoteActiveDoc.Element.SystemName == nil || *remoteActiveDoc.Element.SystemName != *s.resource.Spec.SystemName {
```

Comment:
Does system allow changing the 'SystemName' attribute once the object has been created?

Reply:
No. The systemName always exist. And 3scale API should always return the systemName. That check is because the field is optional in the client library and if for whatever reason systemname is not in the response, the field would be nil.